src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala

```diff
@@ -221,26 +213,82 @@ class Inliner[BT <: BTypes](val btypes: BT) {
+    }
+    private def lst[T](a: Array[T]): java.util.List[T] = java.util.Arrays.asList(a: _*)
+
+    def saveMethodState(methodNode: MethodNode): Unit = if (active && !methodStateSaved) {
+      methodStateSaved = true
+      val currentInstructions = methodNode.instructions.toArray
+      val currentLocalVariables = arr(methodNode.localVariables)
+      val currentTryCatchBlocks = arr(methodNode.tryCatchBlocks)
+      val currentMaxLocals = methodNode.maxLocals
+      val currentMaxStack = methodNode.maxStack
+
+      apply {
+        // this doesn't work: it doesn't reset the `prev` / `next` / `index` of individual instruction nodes
+        // methodNode.instructions.clear()
+        methodNode.instructions.iterator.asScala.toList.foreach(methodNode.instructions.remove)
+        for (i <- currentInstructions) methodNode.instructions.add(i)
+
+        methodNode.localVariables.clear()
+        methodNode.localVariables.addAll(lst(currentLocalVariables))
+
+        methodNode.tryCatchBlocks.clear()
+        methodNode.tryCatchBlocks.addAll(lst(currentTryCatchBlocks))
+
+        methodNode.maxLocals = currentMaxLocals
+        methodNode.maxStack = currentMaxStack
+      }
+    }
+  }
+
+  val NoUndoLogging = new UndoLog(active = false)
 
   /**
    * Inline the callsite of an inlining request and its post-inlining requests.
    *
    * @return An inliner warning for each callsite that could not be inlined.
    */
-  def inline(request: InlineRequest): List[CannotInlineWarning] = canInlineBody(request.callsite) match {
-    case Some(w) =>
-      if (compilerSettings.YoptLogInline.isSetByUser) {
-        val size = request.callsite.callsiteMethod.instructions.size
-        inlineLog ::= InlineLog(request, size, size, 0, Some(w))
-      }
-      List(w)
-    case None =>
+  def inline(request: InlineRequest, undo: UndoLog = NoUndoLogging): List[CannotInlineWarning] = {
+    def doInline(undo: UndoLog): List[CannotInlineWarning] = {
       val sizeBefore = request.callsite.callsiteMethod.instructions.size
-      inlineCallsite(request.callsite)
+      inlineCallsite(request.callsite, undo)
       if (compilerSettings.YoptLogInline.isSetByUser)
         inlineLog ::= InlineLog(request, sizeBefore, request.callsite.callsiteMethod.instructions.size, request.callsite.callee.get.callee.instructions.size, None)
       val postRequests = request.post.flatMap(adaptPostRequestForMainCallsite(_, request.callsite))
-      postRequests flatMap inline
+      postRequests.flatMap(inline(_, undo))
+    }
+
+    canInlineCallsite(request.callsite) match {
+      case None =>
+        doInline(undo)
+
+      case Some((w, illegalAccessInsns)) if illegalAccessInsns.nonEmpty && illegalAccessInsns.forall(ins => request.post.exists(_.callsite.callsiteInstruction == ins)) =>
+        // speculatively inline, roll back if an illegalAccessInsn cannot be eliminated
+        if (undo == NoUndoLogging) {
```

Comment:
Yes. If we start a speculative inline and fail somewhere downstream, it the warning will be pushed up and will cause a rollback.